azure/converters/image_test.go

```diff
@@ -26,6 +27,96 @@ import (
 )
 
 func Test_ImageToPlan(t *testing.T) {
+	cases := []struct {
+		name      string
+		image     *infrav1.Image
+		expect    func(*GomegaWithT, *compute.ImageReference)
+		expectErr bool
+	}{
+		{
+			name: "Should return a specific image reference if image id is specified",
+			image: &infrav1.Image{
+				ID: to.StringPtr("fake-id"),
+				SharedGallery: &infrav1.AzureSharedGalleryImage{
+					SubscriptionID: "fake-sub-id",
+					ResourceGroup:  "fake-rg",
+					Gallery:        "fake-gallery-name",
+					Name:           "fake-image-name",
+					Version:        "v1.0.0",
+				},
+			},
+			expect: func(g *GomegaWithT, result *compute.ImageReference) {
+				g.Expect(result).To(Equal(&compute.ImageReference{
+					ID: to.StringPtr("fake-id"),
+				}))
+			},
+			expectErr: false,
+		},
+		{
+			name: "Should return a market place image reference if Marketplace is specified",
+			image: &infrav1.Image{
+				Marketplace: &infrav1.AzureMarketplaceImage{
+					Publisher: "my-publisher",
+					Offer:     "my-offer",
+					SKU:       "my-sku",
+					Version:   "v1.0.0",
+				},
+			},
+			expect: func(g *GomegaWithT, result *compute.ImageReference) {
+				g.Expect(result).To(Equal(&compute.ImageReference{
+					Publisher: to.StringPtr("my-publisher"),
+					Offer:     to.StringPtr("my-offer"),
+					Sku:       to.StringPtr("my-sku"),
+					Version:   to.StringPtr("v1.0.0"),
+				}))
+			},
+			expectErr: false,
+		},
+		{
+			name: "Should return a SIG image reference if shared gallery is specified",
+			image: &infrav1.Image{
+				SharedGallery: &infrav1.AzureSharedGalleryImage{
+					SubscriptionID: "fake-sub-id",
+					ResourceGroup:  "fake-rg",
+					Gallery:        "fake-gallery-name",
+					Name:           "fake-image-name",
+					Version:        "v1.0.0",
+				},
+			},
+			expect: func(g *GomegaWithT, result *compute.ImageReference) {
+				ID := to.StringPtr(fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/galleries/%s"+
+					"/images/%s/versions/%s", "fake-sub-id", "fake-rg", "fake-gallery-name", "fake-image-name", "v1.0.0"))
+				g.Expect(result.ID).To(Equal(ID))
+			},
+			expectErr: false,
+		},
+		{
+			name:      "Should return error if image is not specified",
+			expectErr: true,
```

Comment:
Wdyt about using a string for the expected error message instead of a bool? We do that [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/inboundnatrules/spec_test.go). The `expectedErr` string let's us check to verify the error message is correct, and we can set it to `""` to represent no error.

```
if tc.expectedError != "" {
	g.Expect(err).To(HaveOccurred())
	g.Expect(err).To(MatchError(tc.expectedError))
} else {
	g.Expect(err).NotTo(HaveOccurred())
	g.Expect(res).To(Equal(tc.expectedPortResult))
}
```